infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/receive/StreamsSnapshotWriter.java

```diff
@@ -304,19 +304,26 @@ private void applyShadowStream(UUID streamId, long snapshot) {
         // as these streams will get trimmed and 'clear' will be a 'data loss'.
         if (MERGE_ONLY_STREAMS.contains(streamId)) {
             log.debug("Do not clear stream={} (merge stream)", streamId);
-        } else if (!replicatedStreamIds.contains(streamId)) {
-            log.trace("No data was written to stream {} on source or sink." +
-                "  Do not clear.", streamId);
-            return;
-        } else {
-            smrEntries.add(CLEAR_ENTRY);
         }
 
+        boolean shouldAddClearRecord = !MERGE_ONLY_STREAMS.contains(streamId);
         while (iterator.hasNext()) {
+            // append a clear record at the beginning of every non-merge-only streams
+            if(shouldAddClearRecord) {
```

Comment:
Minor comment, we can update log message in line 370, to only print: "Start apply shadow streams", without the total field, as it might be misleading cause on snapshot apply resume it will print 0, and it's confusing.